File: 3/3217/DeleteNodesFromLinkedListPresentInArray.java
/*
https://leetcode.com/problems/delete-nodes-from-linked-list-present-in-array/description/?envType=daily-question&envId=2024-09-06
*/

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode modifiedList(int[] nums, ListNode head) {
        boolean alreadySeenFirstValidNode=false;
        ListNode resultantList=new ListNode(nums[0]);
        ListNode tailForResultantList=resultantList;
        HashMap<Integer, Integer>valuesInNumsMapping=new HashMap<>();
        int lengthOfNums=nums.length;
        for(int i=0; i<lengthOfNums; ++i){
            valuesInNumsMapping.put(nums[i], nums[i]);
        }
        while(head!=null){
            int currentHeadValue=head.val;
            if(!valuesInNumsMapping.containsKey(currentHeadValue)){
                if(!alreadySeenFirstValidNode){
                    resultantList=new ListNode(currentHeadValue);
                    tailForResultantList=resultantList;
                    alreadySeenFirstValidNode=true;
                } else{
                    ListNode newNode=new ListNode(currentHeadValue);
                    tailForResultantList.next=newNode;
                    tailForResultantList=tailForResultantList.next;
                }
            }
            head=head.next;
        }
        return resultantList;
    }
}

File: 3/3000s/3217/DeleteNodesFromLinkedListPresentInArray.cpp
/*
https://leetcode.com/problems/delete-nodes-from-linked-list-present-in-array/description/?envType=daily-question&envId=2024-09-06
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* modifiedList(vector<int>& nums, ListNode* head) {
        bool flag=false;
        ListNode *res=new ListNode(head->val);
        map<int, int>mp;
        for(auto &vals: nums){
            mp[vals]++;
        }
        ListNode *tail=res;
        while(head!=nullptr){
            int curval=head->val;
            auto itr=mp.find(curval);
            if(itr==mp.end()){
                if(!flag){
                    res=new ListNode(curval);
                    tail=res;
                    flag=true;
                } else{
                    ListNode *newNode=new ListNode(curval);
                    tail->next=newNode;
                    tail=tail->next;
                }
            }
            head=head->next;
        }
        return res;
    }
};